experimental/proxy_attestation/src/main.rs

```diff
@@ -0,0 +1,168 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Proxy Attestation Service implementation.
+//!
+//! Proxy Attestation Service works as a Certificate Authority for TEE applications. It attests
+//! applications and creates signed certificates for TEE applications, which they use to create TLS
+//! connections with clients.
+//!
+//! The work cycle of the proxy is represented in 2 stages.
+//! First stage corresponds to the backend server attestation:
+//! - TEE application connects to the proxy
+//! - Proxy attests the application using a corresponding remote attestation protocol
+//! - Application sends the proxy a certificate signing request that contains TEE measurements
+//! - Proxy sings it using its private key and sends the signature back to the application
+//!
+//! Second stage corresponds to the client connection:
+//! - Client remotely attests the proxy using a corresponding attestation protocol
+//! - Proxy sends the client its root certificate
+//!     - Client trusts that this is a correct certificate, since it was sent via a secure channel
+//!       created during the attestation process
+//! - Client connects to the application using TLS, and the application uses a TLS certificate
+//!   previously signed by the proxy
+//! - Client checks that the certificate was signed by the root certificate and establishes a secure
+//!   connection
+//!
+//! Note: Current version of Remote Attestation Service doesn't have remote attestation.
+
+mod certificate;
+
+use crate::certificate::CertificateAuthority;
+use anyhow::Context;
+use futures::future::FutureExt;
+use log::info;
+use oak_proxy_attestation::proto::{
+    proxy_attestation_server::{ProxyAttestation, ProxyAttestationServer},
+    GetRootCertificateRequest, GetRootCertificateResponse, GetSignedCertificateRequest,
+    GetSignedCertificateResponse,
+};
+use openssl::x509::X509Req;
+use structopt::StructOpt;
+use tonic::{
+    transport::{Identity, Server, ServerTlsConfig},
+    Request, Response, Status,
+};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Aggregator Backend")]
```

Comment:
Not sure whether these comments are accurate.

Reply:
Done